File: 1501-2000/1905_countSubIslands.py
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        m = len(grid1)
        n = len(grid1[0])
        numIslands = 0
        for i in range(m):
            for j in range(n):
                if grid2[i][j] == 1:
                    self.included = True
                    self.visitIsland(grid1, grid2, m, n, i, j)
                    if self.included:
                        numIslands += 1
        return numIslands

    def visitIsland(self, grid1, grid2, m, n, row, col):
        if m > row >= 0 <= col < n and grid2[row][col]:
            if grid1[row][col] == 0:
                self.included = False
            grid2[row][col] = 0
            self.visitIsland(grid1, grid2, m, n, row - 1, col)
            self.visitIsland(grid1, grid2, m, n, row + 1, col)
            self.visitIsland(grid1, grid2, m, n, row, col - 1)
            self.visitIsland(grid1, grid2, m, n, row, col + 1)

File: 1501-2000/1905_countSubIslands.java
class Solution {
    public int countSubIslands(int[][] grid1, int[][] grid2) {
        int m = grid2.length, n = grid2[0].length;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        int count = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid2[i][j] == 1) {
                    Queue<int[]> queue = new LinkedList<>();
                    queue.add(new int[] {i, j});
                    grid2[i][j] = 0;
                    boolean countable = (grid1[i][j] == 1);

                    while (!queue.isEmpty()) {
                        int[] coordinate = queue.remove();
                        int x = coordinate[0], y = coordinate[1];
                        if (grid1[x][y] == 0) {
                            countable = false;
                        }

                        for (int[] dir: directions) {
                            int r = x + dir[0], c = y + dir[1];
                            
                            if (r >= 0 && r < m && c>= 0 && c < n && grid2[r][c] == 1) {
                                grid2[r][c] = 0;
                                queue.add(new int[] {r, c});
                            }
                        }
                    }

                    count += (countable) ? 1 : 0;
                }
            }
        }

        return count;
    }
}
